lib/linked_list.rb

```diff
@@ -12,120 +12,225 @@ def initialize(value, next_node = nil)
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
-    end
-
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    def add_first(value)
-      raise NotImplementedError
-    end
-
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    def search(value)
-      raise NotImplementedError
-    end
-
-    # method to return the max value in the linked list
-    # returns the data value and not the node
-    def find_max
-      raise NotImplementedError
-    end
-
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    def find_min
-      raise NotImplementedError
-    end
-
-
-    # method that returns the length of the singly linked list
-    def length
-      raise NotImplementedError
-    end
-
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns nil if there are fewer nodes in the linked list than the index value
-    def get_at_index(index)
-      raise NotImplementedError
-    end
-
-    # method to print all the values in the linked list
-    def visit
-      raise NotImplementedError
-    end
-
-    # method to delete the first node found with specified value
-    def delete(value)
-      raise NotImplementedError
-    end
-
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    def reverse
-      raise NotImplementedError
-    end
-
-
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    def find_middle_value
-      raise NotImplementedError
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
+  
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  def add_first(value)
+    @head = Node.new(value, @head)
+  end
+  
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  def search(value)
+    current = @head
+    
+    until current.nil?
+      return true if current.data == value
+      current = current.next
+    end
+    
+    return false
+  end
+  
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    current = @head
+    return nil if current.nil?
+    max = 0

```

Comment:
What if the list only holds negative numbers?

```suggestion
    max = current.data
```